haskellPackages: support template-haskell and test suites when cross compiling

alexfmpe opened the pull request:
Adapted from
https://github.com/input-output-hk/haskell.nix/blob/c0454391b7b026387623289d3deb4ce9b431a528/overlays/linux-cross.nix

~Also unblocks meaningful work on cross testing since TH is used in every other suite for test discovery, including QuickCheck's~ 
~EDIT: Also run test suites ( pkgsCross.ghcjs ones need #451527 )~ Moved to https://github.com/NixOS/nixpkgs/pull/451928/

Building/running `haskell.packages.ghc912.aeson-gadt-th` on x86_64-linux:
- :heavy_check_mark: pkgsStatic
- :heavy_check_mark: pkgsCross.aarch64-multiplatform
- :heavy_check_mark: pkgsCross.aarch64-multiplatform-musl
- :x: pkgsCross.aarch64-android-prebuilt
- :x: pkgsCross.ucrt64

Note `aarch64-android-prebuilt.pkgsStatic` builds for the wrong reasons as 
```
nix-repl> with pkgsCross.aarch64-android-prebuilt.pkgsStatic.stdenv.hostPlatform; { inherit isAarch isMusl isStatic isAndroid useAndroidPrebuilt; }                       
{
  isAarch = true;
  isAndroid = false;
  isMusl = true;
  isStatic = true;
  useAndroidPrebuilt = false;
}
```
which leads to problems down the line (e.g when trying to add system tools). However, trying
```
{
  isAarch = true;
  isAndroid = true;
  isMusl = false;
  isStatic = true;
  useAndroidPrebuilt = true;
}
```
fails when building TH with `iserv-proxy-interpreter: internal error: ASSERTION FAILED: file rts/linker/Elf.c, line 812`
which is https://gitlab.haskell.org/ghc/ghc/-/blob/ghc-9.12.2-release/rts/linker/Elf.c#L812

Example run

```
$ uname -mo
x86_64 GNU/Linux
```
```
$ nix-build -A pkgsCross.aarch64-multiplatform.haskell.packages.ghc912.th-orphans   
...
Running phase: buildPhase
Preprocessing library for th-orphans-0.13.16...
Building library for th-orphans-0.13.16...
[1 of 2] Compiling Language.Haskell.TH.Instances.Internal ( src/Language/Haskell/TH/Instances/Internal.hs, dist/build/Language/Haskell/TH/Instances/Internal.o, dist/build/Language/Haskell/TH/Instances/Internal.dyn_o )
[2 of 2] Compiling Language.Haskell.TH.Instances ( src/Language/Haskell/TH/Instances.hs, dist/build/Language/Haskell/TH/Instances.o, dist/build/Language/Haskell/TH/Instances.dyn_o )
---> Starting interpreter on port 6151
---| interpreter should have started on 6151
Listening on port 6151
---> killing interpreter...
[1 of 2] Compiling Language.Haskell.TH.Instances.Internal ( src/Language/Haskell/TH/Instances/Internal.hs, dist/build/Language/Haskell/TH/Instances/Internal.p_o )
[2 of 2] Compiling Language.Haskell.TH.Instances ( src/Language/Haskell/TH/Instances.hs, dist/build/Language/Haskell/TH/Instances.p_o )
---> Starting interpreter on port 5830
---| interpreter should have started on 5830
Listening on port 5830
---> killing interpreter...
buildPhase completed in 3 minutes 32 seconds
...
``` 
```
> $(nix-build -A qemu-user)/bin/qemu-aarch64 $(nix-build -A pkgsCross.aarch64-multiplatform.haskell.packages.ghc912.aeson-gadt-th)/bin/readme
Encoding of A_a:
"[\"A_a\",[]]"
Decoding of encoded A_a:
Just (Some A_a)

Encoding of (A_b 1):
"[\"A_b\",1]"
Decoding of encoded (A_b 1):
Just (Some (A_b 1))

Encoding of (B_a 'a' (A_b 1)):
"[\"B_a\",[\"a\",[\"A_b\",1]]]"
Decoding of encoded (B_a 'a' (A_b 1)):
Succeeded
```

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alexfmpe committed:
haskellPackages.aeson-gadt-th: re-enable readme executable

alexfmpe committed:
haskellPackages.ghc: disable dynamic linker for android

alexfmpe commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -10,12 +10,79 @@
+    # This is a silly workaround to avoid infinite recursion.
+    # We don't know which packages need template haskell, so we always need iserv-proxy for the host
+    # but then that package also has dependencies which would themselves need iserv-proxy.
+    # TODO: grab host iserv-proxy from the "original" package set without TH support, then have host TH support on the "real" package set.
```
How would I go about doing this without too much pain?

alexfmpe commented:
> Am I holding it wrong?

Hmm no, I am actually.  Here's where we stand

Besides the basic machinery added here, there's a bunch of per-config tweaks that will be needed (e.g. env flags for wine).
I'm mostly been testing with a few `x86_64-linux -> aarch64-linux` because those require minimal tweaking.

https://github.com/NixOS/nixpkgs/blob/5a49a581d1e4be1bb422c3f332ecbfe8d7043dc6/pkgs/development/haskell-modules/generic-builder.nix#L26-L47

The reason I special cased `aarch64-linux` hosts for now is that only relying on `emulatorAvailable` can be a regression, as was recently the case for `pkgsCross.ucrt64` since executables couldn't be built, meaning `iserv-proxy` couldn't be built, meaning *no* packages could be built at all since eval doesn't know ahead of time which need template haskell.


On my `x86_64-linux`, when relaxing the restriction to `isLinux`

- `pkgsStatic.haskell.packages.ghc96.th-orphans` fails as this `libiserv` approach is [> 9.6](https://github.com/NixOS/nixpkgs/pull/445672#discussion_r2373814742)
- `pkgsStatic.haskell.packages.ghc910.th-orphans` fails with some error about gmp/ghc-bignum errors I haven't looked into yet
- `pkgsStatic.haskell.packages.native-bignum.ghc910.th-orphans`  builds
- `pkgsStatic.haskell.packages.native-bignum.ghc912.th-orphans`  builds

After looking at https://github.com/NixOS/nixpkgs/pull/383165, I'd say that it works for `pkgsStatic` *accidentally* because `emulatorAvailable` also looks at `canExecute`. I don't think we need nor want the whole iserv-proxy dance for the static "native-cross".

Seems to me the concerns on my PR and yours interact in subtle ways so I'd say we ought to get at least one of them merged so they can then be addressed together. Plently of layers on this onion to peel anyway.

alexfmpe commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -802,8 +877,19 @@ lib.fix (
 
+      # stdenv.make-derivation sets `doCheck = false` when the build platform can't directly execute the host platform
+      # which means the `checkPhase` ends up disabled and we need to sneak it back in somehow
+      # TODO: avoid this workaround - add some sort of doCheckEmulated?
```
I've been pointed to https://nixos.org/manual/nixpkgs/stable/#var-passthru-tests 
Sounds like a decent compromise?

wolfgangwalther reviewed (commented):
I went commit-by-commit and only did roughly the first half of commits. I think some of the things might be (partly) addressed later already.

Could you try to rebase the branch and make each commit atomic - so that later commits don't introduce changes to code that was only introduced earlier in the same PR, but code is immediately introduced in the way we actually want it? That should make reviewing a lot easier.

wolfgangwalther commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -282,6 +285,55 @@ let
+            (>&2 echo "---> Starting interpreter on port $PORT")
+            ${emulator.path} ${hostProxy} tmp $PORT &
+            (>&2 echo "---| interpreter should have started on $PORT")
+            RISERV_PID="$!"
```
This seems to be still open.

wolfgangwalther commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -10,12 +10,81 @@
+      pname:
+      lib.versionAtLeast ghc.version "9.6"
+      && stdenv.hostPlatform.emulatorAvailable buildPackages
+      && !(crossSupport.iservProxyDependency pname);
```
We're inside the `crossSupport = rec`, so:


```suggestion
      && !(iservProxyDependency pname);
```

alexfmpe replied:
`iservProxyDependency` is now gone

wolfgangwalther commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -10,12 +10,81 @@
+
+    iservWrapper =
+      let
+        buildProxy = iserv-proxy.build + "/bin/iserv-proxy";
```
Refering to the other comment - I don't understand why `buildPackages.iserv-proxy` would be any harder to understand than `iserv-proxy.build`.

We already use `buildPackages` in a lot of places, so why do it differently here? This adds indirection, that I'm not sure we need.

wolfgangwalther commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -10,12 +10,81 @@
+            PORT=$((5000 + $RANDOM % 5000))
+            (>&2 echo "---> Starting interpreter on port $PORT")
+            ${emulator} ${hostProxy} tmp $PORT &
+            (>&2 echo "---| interpreter should have started on $PORT")
```
This log message is pretty useless, I think, because it will never wait for anything that the preceding line does. The `&` detaches, so the echo will happen immediately.. even if the interpreter failed to start.

alexfmpe commented on pkgs/development/tools/haskell/iserv-proxy/default.nix:
```diff
@@ -0,0 +1,56 @@
+  src = fetchFromGitHub {
+    owner = "stable-haskell";
+    repo = "iserv-proxy";
+    rev = "a53c57c9a8d22a66a2f0c4c969e806da03f08c28";
```
Intentionally leaving that for later.
The commit after the pinned one adds a `network` upper bound that rules out our default version. Said bound was added to avoid a bug when running under wine which still needs more work on our end anyway.

alexfmpe commented on pkgs/development/compilers/ghc/common-hadrian.nix:
```diff
@@ -342,7 +342,14 @@ let
+        (targetPlatform.isAndroid || (targetPlatform.isMusl && hostPlatform != targetPlatform))
+        [
+          "*.ghc.cabal.configure.opts += --flags=-dynamic-system-linker"
+        ];
```
I had it as you say but nix-fmt insists in overriding me. 
We can move the comment before `++` but there's no way around this indentation

alexfmpe commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -205,10 +274,15 @@ in
   # this metadata is far from complete in nixpkgs.
   __onlyPropagateKnownPkgConfigModules ? false,
+
+  externalInterpreterTemplateHaskell ? isCross && crossSupport.canProxyTH pname,
```
Done

alexfmpe commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -541,7 +623,12 @@ let
-    exec "$@"
+    ${
+      if (isCross && crossSupport.canCheck pname) then
+        (if stdenv.hostPlatform.isGhcjs then "node" else crossSupport.emulator) + " $@"
```
Made uniform

alexfmpe commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -282,6 +285,55 @@ let
+            RISERV_PID="$!"
+            ${buildProxy} $@ 127.0.0.1 "$PORT"
+            (>&2 echo "---> killing interpreter...")
+            kill $RISERV_PID
```
I tested with and without sandboxing while building from nixos. 
Didn't check building from darwin since when I tried that at the start of this PR, the build of cross ghc itself kept hitting https://gitlab.haskell.org/ghc/ghc/-/issues/24939

sternenseemann commented on pkgs/development/haskell-modules/configuration-ghc-9.6.x.nix:
```diff
@@ -60,6 +59,11 @@ in
   xhtml = null;
   Win32 = null;
 
+  # Was only ever released for a few exact versions of ghc
```
But it's also only required for 9.6? Comment is a little bit confusing to me.

alexfmpe replied:
The whole situation is also confusing to me really: https://github.com/NixOS/nixpkgs/pull/445672#discussion_r2430214570

sternenseemann reviewed (commented):
I'll look at the generic-builder changes tomorrow, too tired today.

wolfgangwalther commented on pkgs/development/tools/haskell/iserv-proxy/default.nix:
```diff
@@ -0,0 +1,53 @@
+{
```
Do we want to be able to recreate this file from cabal2nix in the future? If not, then maybe we can even define the list of non-core dependencies here, provide them as passthru and consume them in configuration-nix for an automatic overlay?

(entirely optional, ofc)

wolfgangwalther commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -541,7 +607,16 @@ let
+            "exec";
+      in
+      runner + " " + ''"$@"''
+    }
```
(optional)

How about:

```suggestion
    exec ${if (isCross && crossSupport.canCheck) then
      (if stdenv.hostPlatform.isGhcjs then "node" else crossSupport.emulator)
    else
      ""} "$@"
```

(not sure what the formatter will do it with it)

This would `exec` in both cases, which might not be bad?

wolfgangwalther reviewed (commented):
Looking great, only some optional nits now.

Not approving, yet, because I would still like to see Sterni's review and would like to run some builds, too.

alexfmpe committed:
haskellPackages: when requested, configure gmp even if cross compiling

Fixes error: "iserv-proxy-interpreter: Failed to lookup symbol:
__gmpn_cmp"

This comes up when building `pkgsStatic.haskell.packages.ghc...`, aka
without `.native-bignum` modifier.

wolfgangwalther commented on pkgs/development/compilers/ghc/common-hadrian.nix:
```diff
@@ -639,7 +639,7 @@ stdenv.mkDerivation (
-    ++ lib.optionals (targetPlatform == hostPlatform && !enableNativeBignum) [
+    ++ lib.optionals (!enableNativeBignum && !targetPlatform.isGhcjs && !targetPlatform.isWasm) [
       "--with-gmp-includes=${targetLibs.gmp.dev}/include"
       "--with-gmp-libraries=${targetLibs.gmp.out}/lib"
```
Doesn't that imply gmp for static? Would that be a problem because of license?

wolfgangwalther replied:
I'll defer to @sternenseemann on this one. I am happy with the remainder of the PR, I think, but am really unsure about the last commit.

sternenseemann replied:
- This doesn't change whether GMP is used or not (this depends solely on `enableNativeBignum`). The point of this condition was probably not to use the `--with-gmp-*` flags if we're building a cross compiler since this can cause to incorrect paths or `-L` flags being used by Hadrian since it doesn't actually distinguish between host and target gmp. Instead the idea is to just let the cc wrapper figure it out. The downside of this is usually that the package db entries don't have the necessary paths. Would need some blaming to figure out why this was added originally.
- @alexfmpe We did you end up enabling the flag again? Does this solve a specific problem?
- The `targetPlatform.isGhcjs` is redundant since we already flick on `enableNativeBignum` for that. We could do the same for wasm, but I don't entirely see the point since it's purely theoretical.

```suggestion
    ++ lib.optionals (!enableNativeBignum) [
      "--with-gmp-includes=${targetLibs.gmp.dev}/include"
      "--with-gmp-libraries=${targetLibs.gmp.out}/lib"
```

sternenseemann commented on pkgs/development/tools/haskell/iserv-proxy/default.nix:
```diff
@@ -0,0 +1,56 @@
+  pname = "iserv-proxy";
+  version = "9.3";
+
+  # https://github.com/stable-haskell/iserv-proxy/pull/1
```
Should probably talk to the stable haskell/haskell.nix folks what the state is there, but given that haskell.nix depends on it in some capacity, I'm not worried it will go away suddenly?

alexfmpe commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -10,12 +10,65 @@
+          in
+          buildPackages.writeShellScriptBin ("iserv-wrapper" + lib.optionalString enableProfiling "-prof") ''
+            set -euo pipefail
+            PORT=$((5000 + $RANDOM % 5000))
```
Yeah leaving it for follow-up

sternenseemann commented:
> Oh, keep in mind this all works with 9.10.

Aren't there some conditionals gating on >= 9.12?

sternenseemann commented:
LGTM, but I agree with @wolfgangwalther that the `iserv-proxy = { … }` argument is awkward and best redesigned. I think I have an idea on how to do it. In principle, I'd be fine with merging this, and I try to implement that in a follow up. We are going to be on unstable for a while now, so we can afford to do that :)

WDYT @wolfgangwalther?

alexfmpe commented:
I just noticed this introduces two regressions on `pkgsCross.ucrt64.haskell.packages.ghc912.hello`.
Pushing a few simple commits to fix that

wolfgangwalther commented on pkgs/top-level/release-haskell.nix:
```diff
@@ -520,7 +520,26 @@ let
+                  hello
+                  microlens
+                  miso
+                  reflex-dom
```
`reflex-dom` currently doesn't eval because libsoup is marked as insecure. We should probably remove it from here.

wolfgangwalther replied:
I removed `reflex-dom` here and also used the default `haskellPackages` for the release-haskell file cross to aarch64 thing instead of pinning 9.12.

wolfgangwalther committed:
WIP: haskellPackages: support Template Haskell when cross compiling

These changes were split from the commit before, keeping for future
reference for now.

wolfgangwalther commented on pkgs/development/compilers/ghc/common-hadrian.nix:
```diff
@@ -340,7 +340,14 @@ let
+      lib.optionals
+        (targetPlatform.isAndroid || (targetPlatform.isMusl && hostPlatform != targetPlatform))
+        [
+          "*.ghc.cabal.configure.opts += --flags=-dynamic-system-linker"
```
The condition here is wrong, according to my experiments. Cross with musl works fine, but once we do `isStatic` it doesn't. Quite obviously.

I'm not sure whether we should just check `isStatic` - or whether we should actually check `enableShared`, which includes `isStatic` and more.

Given that the error message for android is:

```
qemu-aarch64: Could not open '/system/bin/linker64': No such file or directory
```

... I wonder whether we should actually add the `isAndroid` condition to `enableShared` to begin with?

wolfgangwalther reviewed (commented):
I worked through this PR and rebased. Here are some notes on what I did and noticed:
- I rebased the branch on very latest haskell-updates, which I bumped hackage on ~~today~~ yesterday.
- The last commit was to bump `iserv-proxy` - I squashed that into the first commit and also adjusted the version number with the correct unstable date.
- At the first commit, I can build `iserv-proxy` for GHC 9.8, 9.10 and 9.12 - which seems reasonable: The default version + one version in both directions. I can't build `iserv-proxy` for GHC 9.4, GHC 9.6 or GHC HEAD, the latter fails somewhere in the dependencies.
- The second commit contained changes specifically for GHC 9.6 and libiserv. But `iserv-proxy` does not build with or without these changes for GHC 9.6, so I backed these out and put them into a fixup to look at later, if at all. We can be happy to be able to do this for GHC 9.8+, I think.
- The second commit also contained changes to use the external interpreter for test suites. This contained the following line `exec ${if (isCross && crossSupport.canCheck) then "node" else crossSupport.emulator} "$@"`. This seems wrong, because why should we always use `node` when in a cross setting - and if not fallback to the cross emulator? I think the condition should rather be on `isGhcjs` or so? I removed this for now, to reduce scope a bit - and to prevent rebuild of the default non-cross compiler. This can be introduced in a follow up again, to enable testing with the external interpreter - let's focus on TH as the first step. I kept the relevant changes as a fixup commit for future reference.
- With the second commit, I am able to build `th-orphans`, which was not possible in any cross setting before:
  - aarch64 -> x86_64 (gnu)
  - aarch64 -> x86_64 (musl)
  - x86_64 -> aarch64 (gnu)
  - (did not test musl for x86 -> aarch64)
- Notably the things that didn't work on the second commit were:
  - `pkgsStatic` - which failed with `error: Dynamic loading not supported` - I believe this is fixed by the third commit, but the conditions (`android || musl`) seem wrong. I think this should be conditioned on `isStatic`.
  - aarch64-darwin fails to build the compiler. Haven't looked deeper into that, yet.
  - `pkgsCross.aarch64-android-prebuilt` failed with: `qemu-aarch64: Could not open '/system/bin/linker64': No such file or directory`.

---

I just pushed my current state. I will let some builds finish and then plan to push *the first few commits* to haskell-updates already, but not merge the full PR, yet. We can then iterate on the remaining commits, where I still have question marks for me.

wolfgangwalther commented on pkgs/development/haskell-modules/configuration-cross.nix:
I removed `configuration-cross.nix` - and put the additional packages with disabled external interpreter into `configuration-windows.nix` instead, since they only come up as dependencies of `network` on that platform.

wolfgangwalther reviewed (commented):
I tested more stuff again and I don't think any of the remaining bits are ready to merge right now.

How should we proceed - would you like to take these and open separate PRs for them, with more detailed reasoning and showing how they help? Or should we keep them together in this PR?

alexfmpe commented on pkgs/development/haskell-modules/configuration-nix.nix:
```diff
@@ -1259,9 +1259,6 @@ builtins.intersectAttrs super {
 
-  # Compiling the readme throws errors and has no purpose in nixpkgs
-  aeson-gadt-th = disableCabalFlag "build-readme" super.aeson-gadt-th;
-
```
This is one package I know of where *the executable* uses TH, so I was using it as a sanity check. 
Haven't found a case where TH libs work but an executable doesn't so I don't feel strongly about it.

alexfmpe commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -541,7 +607,7 @@ let
     fi
 
-    exec "$@"
+    exec ${if (isCross && crossSupport.canCheck) then "node" else crossSupport.emulator} "$@"
```
Huh yeah stuff got lost in the last rebase

alexfmpe replied:
https://github.com/NixOS/nixpkgs/pull/468007

alexfmpe commented on pkgs/development/compilers/ghc/common-hadrian.nix:
```diff
@@ -342,7 +342,7 @@ let
     ]
     # Inform GHC that we can't load dynamic libraries which forces iserv-proxy to load static libraries.
-    ++ lib.optionals targetPlatform.isStatic [
+    ++ lib.optionals (targetPlatform.isAndroid || targetPlatform.isStatic) [
```
> qemu-aarch64: Could not open '/system/bin/linker64': No such file or directory

That happens with any android build that requires the dynamic linker, unless you run it on an android system
```
$ NIXPKGS_ALLOW_UNFREE=1 nix-build -A pkgsCross.aarch64-android-prebuilt.hello
/nix/store/f60sv415dxv7am0f06lm956jn2ni4nja-hello-aarch64-unknown-linux-android-2.12.2

$ file result/bin/hello
result/bin/hello: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /system/bin/linker64, not stripped

$ ./result-qemu-user/bin/qemu-aarch64 result/bin/hello
qemu-aarch64: Could not open '/system/bin/linker64': No such file or directory

$ NIXPKGS_ALLOW_UNFREE=1 nix-build -A pkgsCross.aarch64-android-prebuilt.pkgsStatic.hello
/nix/store/5jzni4r3gyi0aj7hbx9w39xwn9c523sh-hello-static-aarch64-unknown-linux-musl-2.12.2

$ file result/bin/hello
result/bin/hello: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, not stripped

$ ./result-qemu-user/bin/qemu-aarch64 result/bin/hello
Hello, world!
```

> What did you do that made the isAndroid condition here an improvement?

IIRC it made TH work for `pkgsCross.aarch64-android-prebuilt.pkgsStatic` but that config is a strange creature, as stated in the PR description.

> GHC 9.8 fails because it can't apply an android related patch, I doubt you had been testing with that before.

Yeah, I only got android to work on 9.10+. I think something on GHC configure would need backporting for 9.8

alexfmpe commented on pkgs/development/compilers/ghc/common-hadrian.nix:
```diff
@@ -636,7 +636,7 @@ stdenv.mkDerivation (
       "--with-ffi-libraries=${targetLibs.libffi.out}/lib"
     ]
-    ++ lib.optionals (targetPlatform == hostPlatform && !enableNativeBignum) [
+    ++ lib.optionals (!enableNativeBignum) [
```
https://github.com/NixOS/nixpkgs/pull/481162

wolfgangwalther commented:
> * The second commit contained changes specifically for GHC 9.6 and libiserv. But `iserv-proxy` does not build with or without these changes for GHC 9.6, so I backed these out and put them into a fixup to look at later

Just noted that the upstream iserv-proxy PR now contains https://github.com/stable-haskell/iserv-proxy/pull/1/changes/dc144516464d3b4f4107b1765873473758aae1dc..91ef7ffdeedfb141a4d69dcf9e550abe3e1160c6, which seems to fix something for GHC < 9.7, so it might be worth looking into 9.6 again here.
